### PR TITLE
add web server

### PR DESCRIPTION
**Pull request by kungyao**

### CODE REVIEW

File: src/WebServer.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+// found in the LICENSE file.
+
+#ifndef HTTPLIB_SRC_WEBSERVER_H
+#define HTTPLIB_SRC_WEBSERVER_H
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#define DEFAULT_IP "127.0.0.1"
```
Comment (john0312) [inline src/WebServer.h:12]: Please leave a TODO to allow configuration of IP and port.

Comment (john0312) [reply src/WebServer.h:12]: Please make it a static constexpr member of your class.

File: src/WebServer.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#define DEFAULT_PORT 80
+// 
+class HttpListener;
+class SocketSession;
+
+/** @brief Manage all web-related jobs. */
+class WebServer {
+private:
+    /** @brief Manage all web-related jobs. */
+    static std::shared_ptr<WebServer> web_;
```
Comment (john0312) [inline src/WebServer.h:22]: Please add Get() for singleton.
Consider using raw pointer for singleton.
Please add comment on this being a singleton.

File: src/WebServer.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    /** @brief Request listener. */
+    std::shared_ptr<HttpListener> listener_;
```
Comment (john0312) [inline src/WebServer.h:24]: Please consider using unique_ptr

File: src/WebServer.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    std::vector<std::shared_ptr<SocketSession>> sockets_;
+public:
+    WebServer(const std::string& ip = DEFAULT_IP, const int& port = DEFAULT_PORT);
+    WebServer(const WebServer&) = delete;
+
+    /** @brief Project runner. */
+    void run();
+};
+
+std::shared_ptr<WebServer> WebServer::web_ = nullptr;
```
Comment (john0312) [inline src/WebServer.h:35]: Please move this to .cc file.

File: src/WebServer.h (hunk 1 above)
Comment (kungyao) [reply src/WebServer.h]: done

File: src/WebServer.h (hunk 3 above)
Comment (kungyao) [reply src/WebServer.h]: done

File: src/WebServer.h (hunk 4 above)
Comment (kungyao) [reply src/WebServer.h]: done

File: src/WebServer.h (hunk 2 above)
Comment (kungyao) [reply src/WebServer.h]: we will add a manager class between document and socket later, so we remove class instance here.

**Review by john0312 — APPROVED** [review]

File: src/WebServer.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+private:
+    static constexpr char DEFAULT_IP [] = "127.0.0.1";
+    static constexpr int DEFAULT_PORT = 80;
+private:
+    /** @brief Request listener. */
+    std::unique_ptr<HttpListener> listener_;
+public:
+    WebServer(const std::string& ip = WebServer::DEFAULT_IP, const int& port = WebServer::DEFAULT_PORT);
+    WebServer(const WebServer&) = delete;
+
```
Comment (john0312) [inline src/WebServer.h:24]: Is this copyable?

Comment (kungyao) [reply src/WebServer.h]: done

**Review by john0312 — APPROVED** [review]
Approve assuming naming is changed.

**Pull request merged**